### PR TITLE
add jupyter_contrib_nbextensions recipe

### PR DESCRIPTION
**Pull request by jcb91**
As discussed in #786, this should hopefully build the main repo for `ipython-contrib/IPython-notebook-extensions` (which has now been renamed `jupyter_cntrib_nbextensions`)

### CODE REVIEW

**Comment by jakirkham** [comment]
Seeing the following error in the testing phase on Travis CI.

```
Enabling: jupyter_nbextensions_configurator
- Writing config: /Users/travis/miniconda/envs/_test/etc/jupyter
    - Validating...
      jupyter_nbextensions_configurator  OK
Enabling notebook extension nbextensions_configurator/config_menu/main...
Enabling tree extension nbextensions_configurator/tree_tab/main...
Error: post-link failed for: local::jupyter_contrib_nbextensions-0.1.0-py27_0
```

**Comment by jcb91** [comment]
@jakirkham yeah, I saw that, but I'm so far none the wiser as to what it's actually referring to as an error. I'll investigate today, hopefully...

**Comment by jcb91** [comment]
ok, I'm in a massive muddle now, getting the same error in my local build, even having removed the `post-link.sh` script entirely 😕

**Comment by jcb91** [comment]
mm, right, seems like the problem is with the `test` section (which is part of post-link phase), rather than post-link script. Still no further information on what the error is though...

File: recipes/jupyter_contrib_nbextensions/post-link.bat [hunk 1]
```diff
@@ -0,0 +1,5 @@
+@echo off
+:: We redirect stderr & stdout to conda's .messages.txt; for details, see
+::     http://conda.pydata.org/docs/building/build-scripts.html
+"%PREFIX%\Scripts\jupyter.exe" contrib nbextension install --sys-prefix >> "%PREFIX%/.messages.txt" 2>&1
```
Comment (jakirkham) [inline recipes/jupyter_contrib_nbextensions/post-link.bat]: Maybe `contrib nbextension` -> `contrib_nbextension` here and for the rest of them?

File: recipes/jupyter_contrib_nbextensions/pre-unlink.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# We redirect stderr & stdout to conda's .messages.txt; for details, see
+# http://conda.pydata.org/docs/building/build-scripts.html
+"${PREFIX}/bin/jupyter" nbextension uninstall --sys-prefix >> "${PREFIX}/.messages.txt" 2>&1
```
Comment (jakirkham) [inline recipes/jupyter_contrib_nbextensions/pre-unlink.sh]: Guessing `nbextension` -> `contrib_nbextension`.

File: recipes/jupyter_contrib_nbextensions/post-link.bat (hunk 1 above)
Comment (jcb91) [inline recipes/jupyter_contrib_nbextensions/post-link.bat]: no, that is how it's meant to be called, and it works in CI testing (see, e.g. https://travis-ci.org/ipython-contrib/jupyter_contrib_nbextensions/jobs/148365017#L395 from https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/master/tests/test_application.py#L135

File: recipes/jupyter_contrib_nbextensions/pre-unlink.sh (hunk 2 above)
Comment (jcb91) [inline recipes/jupyter_contrib_nbextensions/pre-unlink.sh]: this, however is a mistake: it should be `contrib nbextension`

File: recipes/jupyter_contrib_nbextensions/post-link.bat (hunk 1 above)
Comment (jakirkham) [inline recipes/jupyter_contrib_nbextensions/post-link.bat]: Alright, sorry, was just guessing after looking at code in other extension packages for clues.

Comment (jakirkham) [inline recipes/jupyter_contrib_nbextensions/post-link.bat]: Do we need to enable as well?

**Comment by jakirkham** [comment]
Yeah, I'm a little worried we forgot the enabled/disabled steps and something might be lingering from the build environment causing issues when installed in the test environment.

Also, @bollwyvl recommended to me to do the install step as part of the build. Then one doesn't need an uninstall step. With that change one just does enable/disable steps in the link scripts. Seems like sound advice and it has worked for me in the past. Maybe give that a try?

Anyways I'm going to be out for a week. So I'll try to help take another look when I get back (assuming you haven't already figured it out by then).

**Comment by bollwyvl** [comment]
Looks like the approach conflates the notebook upstream's `install` (which only adds files to `/share/`) and `enable` (which modifies files in `/etc/`), so the post/pre steps are necessary.

Still looking around...

**Comment by bollwyvl** [comment]
Building locally, i see this:

```bash
$ more ~/miniconda3/envs/_test/.messages.txt
Enabling: jupyter_nbextensions_configurator
- Writing config: /Users/nbollweg/miniconda3/envs/_test/etc/jupyter
    - Validating...
      jupyter_nbextensions_configurator  ESC[32mOKESC[0m
Enabling notebook extension nbextensions_configurator/config_menu/main...
Enabling tree extension nbextensions_configurator/tree_tab/main...
[JupyterContribApp] CRITICAL | Bad config encountered during initialization:
[JupyterContribApp] CRITICAL | Unrecognized flag: '--sys-prefix'
community-contributed spice for Jupyter Interactive Computing

Options
-------

Arguments that take values are actually convenience aliases to full
Configurables, whose aliases are listed on the help line. For more information
on full configurables, see '--help-all'.

--generate-config
    generate default config file
-y
    Answer yes to any questions instead of prompting.
--debug
    set log level to logging.DEBUG (maximize logging output)
--config=<Unicode> (JupyterApp.config_file)
    Default: ''
    Full path of a config file.
--log-level=<Enum> (Application.log_level)
    Default: 30
    Choices: (0, 10, 20, 30, 40, 50, 'DEBUG', 'INFO', 'WARN', 'ERROR', 'CRITICAL')
    Set the log level by value or name.

To see all available configurables, use `--help-all`
```

**Comment by jcb91** [comment]
@bollwyvl thanks for that, for some reason it didn't occur to me to actually look at `.messages.txt`, having redirected output there.

>  @bollwyvl recommended to me to do the install step as part of the build

Ah, ok, I guess we could do that, although that'd mean all the nbextensions would stay installed on uninstalling the package. The idea here was essentially to provide a wrapper around the upstream `jupyter nbextension install`, but to leave all of the `enable` stuff to the user (the repo has too many extensions to make enabling them all appropriate). Do you think it'd be more appropriate to adopt the simpler in-build-step approach?

**Comment by bollwyvl** [comment]
> although that'd mean all the nbextensions would stay installed on uninstalling the package

conda just links the files... not worth digging too hard into it here, but suffice to say, whatever is in your `.tar.bz2` will be unioned with whatever came before it. At uninstall, it gets unlinked. it's a bummer to ship the files twice (once in `/lib/python` and once in `/share/`), but bzip does wonders for perfect copies of files, so don't bother yourself too much about that, i guess.

> The idea here was essentially to provide a wrapper around the upstream jupyter nbextension install, but to leave all of the enable stuff to the user (the repo has too many extensions to make enabling them all appropriate). 

But I don't think you'll lose that... they'll just already be in `/share/jupyter/nbextensions`, so you're still just manipulating the one (plus _n_) race condition files in `/etc/`. So if you do in fact have to `enable` soemthing, that has to be in `post-pre-un-link.bat.sh`. If you're just copying files, it can go in `build.bld.bat.sh`... or, better still, right in the `meta.yml`, if it doesn't touch paths!

By the by, I love what you're doing. I don't like the install/enable distinction, but that's not changing anytime soon. I have argued for a metadata-based approach like the one you have, but have been shot down on a few occasions. I think it's maybe a bit unfortunate that you selected a new format vs reusing an existing one, even though I love yml... it's not going to get a lot of uptake, as there are some opponents of it.

> Do you think it'd be more appropriate to adopt the simpler in-build-step approach?

see above!

**Comment by jcb91** [comment]
Thanks very much @bollwyvl 😊 

> At uninstall, it gets unlinked

Ah, I didn't realise that, it's neat (though, now that I think about it, it makes perfect sense that this would happen!). This would be a much better option then, _if_ we didn't need to edit any config files, but, now that I think about it more, although we don't enable any extensions, we do edit the `jupyter_nbconvert_config.json`, so I guess we still need to run this after the build is finished (correct me if I've misunderstood!).

> I don't like the install/enable distinction, but that's not changing anytime soon

yeah, it makes naming things like our wrapper script a little tricky to get right! Personally, I quite liked the approach I've adopted in `jupyter_nbextensions_configurator`, where the server's nbextensions path is edited by a serverextension at runtime to include the nbextension files' location. It made sense in that case, for nbextensions which only work when the serverextension is enabled, but adding yet another approach to installing pure nbextensions was a recipe for more confusion in the end, so we went with a wrapper around upstream jupyter machinery.

> I think it's maybe a bit unfortunate that you selected a new format vs reusing an existing one, even though I love yml

Yeah, I did suggest switching to json, which would have been easier to integrate with the npm move, but yaml won out on ease of editing for newcomers. I think I'd like to see `jupyter_nbextensions_configurator` support using package.json or something similar in addition to the current yaml method, so if you have any thoughts on how best that might work, do let me know :)

**Comment by bollwyvl** [comment]
>  so I guess we still need to run this after the build is finished (correct me if I've misunderstood!).

Yeah, but if you could break it up with switches... that's a lot of `copying x ... x` output, vs "look, i've changed nbconvert a little". Anyhow, we should _try_ to do as little as possible in the link scripts... contending for those files is death, and copying is slow, while hardlinks are fast and reliable. Presumably at some point we'll have a reasonable conf.d style setup, but there are apparently still issues around it.

> so we went with a wrapper around upstream jupyter machinery.

Well, props to that... but we're really trying to get away from every extension author having to write a whole suite of stuff just to get things installed. Honestly, the user doesn't care, as long as it works like they expect it to, but I hate maintaining the "boilerplate" around these things. 

> nbextensions path is edited by a serverextension

Nah, i don't like that as much. I mean, go for it, but we have _enough_ locations where your a user's assets might actually be. `nbextension list` is handy to show if one has conflicting installs (user always wins)... as it is, everything installed this way throws the ol' fail ❌ .

> switch to use sha256 checksum

Hooray!

**Comment by jcb91** [comment]
>Yeah, but if you could break it up with switches... that's a lot of copying x ... x output, vs "look, i've changed nbconvert a little".

True. So we should have, e.g. an `install --config-only` and an  `install --no-config`, for example? Sounds reasonable...

> Honestly, the user doesn't care, as long as it works like they expect it to, but I hate maintaining the "boilerplate" around these things.

Absolutely. Since it's all already live, I've seen multiple 'solutions' to this by specifying pypi dependencies and the like, but having multiple alternatives means none really becomes a de-facto standard. I've found myself thinking "I could turn these nbextension install scripts into classes that everyone could inherit from in `jupyter_contrib_core`", then realising that it's pretty me much reinventing yet another different colour of wheel, so not very useful. All power to anything that fixes the boilerplating :)

> everything installed this way throws the ol' fail ❌.

Haha yeah, I had to fudge enabling the nbextensions to get around this, so yeah, I absolutely concede it's not the greatest solution!

**Comment by jcb91** [comment]
So, I think this should work ok now that windows builds of `jupyter_contrib_core` 0.3 have completed, but I can't restart the appveyor build, so I'm just gonna amend the last commit to get it re-triggered...

**Comment by jcb91** [comment]
yay! 🎉 🎈

**Comment by bollwyvl** [comment]
huzzah!

**Comment by ocefpaf** [comment]
Thanks @jcb91! That is a nice addition!!

**Comment by jakirkham** [comment]
Nice work guys. Thanks @bollwyvl for helping @jcb91 get this one working.

**Pull request merged**